### DISCUSSION TITLE
Why does the HRRR model have 2-dimensional grids for latitude and longitude?

### DISCUSSION THREAD

**Discussion by WToma** [blaylockbk/Herbie#45] [Q&A] [answered]
After reading https://cfconventions.org/Data/cf-conventions/cf-conventions-1.1/build/ch05s02.html and several [Stackoverflow](https://stackoverflow.com/questions/63169963/why-latitudes-and-longitudes-are-two-dimensional-arrays-in-netcdf-file) [answers](https://stackoverflow.com/questions/47191930/subsetting-2-d-latitudes-longitudes-and-data-from-a-big-dataset-in-python-ims) I'm still unclear how to properly interpret the situation when the `latitude` and `longitude` coordinates are 2D themselves.

Following the example on the [Herbie main page](https://blaylockbk.github.io/Herbie/_build/html/index.html): 
```
>>> H = Herbie('2021-01-01 12:00', model='hrrr', product='sfc', fxx=6)
>>> H.download(':500 mb')
>>> x = H.xarray('TMP:2 m')
>>> x.coords['latitude']
<xarray.DataArray 'latitude' (y: 1059, x: 1799)>
array([[21.138123, 21.14511 , 21.15209 , ..., 21.154509, 21.147531, 21.140547],
       [21.162995, 21.169984, 21.176967, ..., 21.179387, 21.172407, 21.165419],
       [21.187869, 21.194861, 21.201847, ..., 21.204268, 21.197285, 21.190294],
       ...,
       [47.789559, 47.799849, 47.810129, ..., 47.813691, 47.803415, 47.793128],
       [47.814093, 47.824386, 47.834669, ..., 47.838233, 47.827953, 47.817664],
       [47.838623, 47.84892 , 47.859206, ..., 47.862771, 47.852488, 47.842195]])
Coordinates:
    time               datetime64[ns] 2021-01-01T12:00:00
    step               timedelta64[ns] 06:00:00
    heightAboveGround  float64 2.0
    latitude           (y, x) float64 21.14 21.15 21.15 ... 47.86 47.85 47.84
    longitude          (y, x) float64 ...
    valid_time         datetime64[ns] 2021-01-01T18:00:00
Dimensions without coordinates: y, x
Attributes:
    units:          degrees_north
    standard_name:  latitude
    long_name:      latitude
# similar for longitude, but the values are different
```

It looks like both latitude and longitude describe the whole grid each just by themselves, and two grids are different.

Trying to make sense of this, I looked at `pluck_points` (it's easier for me to understand than `nearest_points`): 
https://github.com/blaylockbk/Herbie/blob/f1ae91b54e827d26a06692bd6f188313be7a097c/hrrrb/tools.py#L132-L143

The way I'm reading this is: "the location you're searching for may be best matched either by a point in `latitide`, or by a point in `longitude`, or by point wherein the lat is in `latitude` and the long is in `longitude`, or the other way". In other words, `latitude` and `longitude` are both full grids in and of themselves, we might as well just call them `grid1` and `grid2`. If we entirely disregard one of them, we still have complete information. I'm extremely suspicious of this interpretation, but it's the only one that makes any sense in my head so far :D

So with that, the questions are
- what's the best way to think of these 2D latitude / longitude coordinates?
- is there a rule, or rule of thumb, about when should I use one or the other?
- ... or are they OK to mix and match, like in my interpretation above?

Thank you!

**Comment by blaylockbk** [comment, marked as answer]
Many limited-area NWP models (the HRRR included) do not have a [Cartesian](https://en.wikipedia.org/wiki/Equirectangular_projection) coordinate system. The HRRR model domain is on a [lambert-conformal](https://en.wikipedia.org/wiki/Lambert_conformal_conic_projection) projection, and 2D coordinates for both lat and lon are required to describe the location of each model point on the earth. The alternative is to use the projection's coordinate system, but that is not as intuitive as lat/lon coordinates.

1) The best way to think of 2D lat/lon is to understand that they are on projected coordinates.
2) If you want to locate the data on a map, you need to know _both_ 2D lat/lon coordinates.

For what it's worth, plucking out nearest-neighbor points for curvilinear grids is not trivial; xarray has been trying to add this capability for several years. My `pluck_points` method you were looking at is illustrated here: https://github.com/blaylockbk/pyBKB_v3/blob/master/demo/Nearest_lat-lon_Grid.ipynb. That might help you see how it finds the nearest point in a grid.

**Reply by WToma** [reply]
Thanks! I think I get it now.

So for example the values of the array `latitude[0]` plots the first line in the model's grid. This does not numerically correspond to a parallel, so in an the equirectangular projection it would not appear as a straight line. However presumably it would appear as a straight horizontal line in the Lambert projection?

**Reply by blaylockbk** [reply]
That is exactly right!

Here is how the HRRR 2-m temperatures looks on an equirectangular (PlateCarree) map versus a Lambert-conformal map; same data, different map projections. Notice how the parallels curve in the second plot.

> I highlighted rows 600-650 in grey (`ds.isel(y=slice(600-650)`)

![image](https://user-images.githubusercontent.com/6249613/153904743-d83f8287-26d5-4320-a79c-92bac572f737.png)
![image](https://user-images.githubusercontent.com/6249613/153904764-18af5c4f-3744-400c-bd71-187e0fb379a0.png)

**Reply by blaylockbk** [reply]
Thanks for the good question.

**Reply by williamhobbs** [reply]
@blaylockbk, this might be a basic (and slightly off-topic) question, but how did you get the latitude and longitude values to show up on the plot axes in the examples above? 

It seems that something like

```
ax.set_xticks(ds.longitude)
ax.set_yticks(ds.latitude)
```

would work, but `latitude` and `longitude` are 2D with HRRR.

**Reply by blaylockbk** [reply]
Does this answer your question? https://scitools.org.uk/cartopy/docs/latest/matplotlib/gridliner.html

**Reply by williamhobbs** [reply]
Yes! Thanks!

**Comment by maxdow** [comment]
Thanks for this helpful informations :) 
Is there a way to convert this HRRR grid to a regular cartesian grid ? I can't find a way to do this but i'm sure it is possible with xarray ?

I try to achieve this kind of indexing ( example from Icon dataset ) 

![image](https://github.com/blaylockbk/Herbie/assets/711050/b57ba40d-1b46-4926-a5f4-37177d0ed47e)

**Reply by blaylockbk** [reply]
As far as I understand, the only way to convert the lat/lon to a cartesian grid is to transform the lat/lon points to the HRRR map projection coordinates (lambert conic conformal). You can look into cartopy and metpy for methods to do that.

I've done something like this before in Herbie here https://github.com/blaylockbk/Herbie/blob/a45da720e2c1a94a64b5b86e4b39a8c3175c7dd2/herbie/accessors.py#L262-L276

**Reply by maxdow** [reply]
thanks for your feedback. 
I tested this approach with pyproj

```python
import cartopy.crs as ccrs

projection = ccrs.LambertConformal(
    central_longitude=262.5,
    central_latitude=38.5,
    standard_parallels=(38.5, 38.5),
    globe=ccrs.Globe(semimajor_axis=6371229, semiminor_axis=6371229),
)
transformer = pyproj.Transformer.from_crs(projection.to_wkt(),4326,always_xy=True)

xv, yv = np.meshgrid(ds.x, ds.y)
lon, lat = transformer.transform(xv, yv)
```

I will try to figure out how to change the dimensions on the grid according to this projection
